# Assignment 5 - Text Analysis
An explanation this assignment could be found in the .pdf explanation document


## Materials to review for this assignment
<h4>From Moodle:</h4> 
<h5><u>Review the notebooks regarding the following python topics</u>:</h5>
<div class="alert alert-info">
&#x2714; <b>Working with strings</b> (tutorial notebook)<br/>
&#x2714; <b>Text Analysis</b> (tutorial notebook)<br/>
&#x2714; <b>Hebrew text analysis tools (tokenizer, wordnet)</b> (moodle example)<br/>
&#x2714; <b>(brief review) All previous notebooks</b><br/>
</div> 
<h5><u>Review the presentations regarding the following topics</u>:</h5>
<div class="alert alert-info">
&#x2714; <b>Text Analysis</b> (lecture presentation)<br/>
&#x2714; <b>(brief review) All other presentations</b><br/>
</div>

## Preceding Step - import modules (packages)
This step is necessary in order to use external modules (packages). <br/>

In [101]:
# --------------------------------------
import pandas as pd
import numpy as np
# --------------------------------------


# --------------------------------------
# ------------- visualizations:
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
# --------------------------------------


# ---------------------------------------
import sklearn
from sklearn import preprocessing, metrics, pipeline, model_selection, feature_extraction 
from sklearn import naive_bayes, linear_model, svm, neural_network, neighbors, tree
from sklearn import decomposition, cluster

from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV 
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.metrics import precision_score, recall_score, f1_score
from sklearn.metrics import mean_squared_error, r2_score, silhouette_score
from sklearn.preprocessing import MinMaxScaler, StandardScaler, LabelEncoder

from sklearn.svm import LinearSVC
from sklearn.neural_network import MLPClassifier
from sklearn.linear_model import Perceptron, SGDClassifier
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.naive_bayes import MultinomialNB, GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
# ---------------------------------------

# ----------------- output and visualizations: 
import warnings
from sklearn.exceptions import ConvergenceWarning
warnings.simplefilter("ignore")
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter("ignore", category=ConvergenceWarning)
# show several prints in one cell. This will allow us to condence every trick in one cell.
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
%matplotlib inline
pd.pandas.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)
# ---------------------------------------

### Text analysis and String manipulation imports:

In [102]:
# --------------------------------------
# --------- Text analysis and Hebrew text analysis imports:
# vectorizers:
from sklearn.feature_extraction import text
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# regular expressions:
import re
# --------------------------------------

### (optional) Hebrew text analysis - WordNet (for Hebrew)
Note: the WordNet is not a must

#### (optional) Only if you didn't install Wordnet (for Hebrew) use:

In [3]:
# word net installation:

# unmark if you want to use and need to install
!pip install wn
!python -m wn download omw-he:1.4



Download (0 bytes) Requesting

Download [                              ] (0/315276 bytes) Receiving

Download [=                             ] (8192/315276 bytes) Receiving

Download [#-                            ] (16384/315276 bytes) Receiving

Download [##-                           ] (24576/315276 bytes) Receiving

Download [###                           ] (32768/315276 bytes) Receiving

Download [###=                          ] (40960/315276 bytes) Receiving

Download [####=                         ] (49152/315276 bytes) Receiving

Download [#####-                        ] (57344/315276 bytes) Receiving

Download [######                        ] (65536/315276 bytes) Receiving

Download [#######                       ] (73728/315276 bytes) Receiving

Download [#######=                      ] (81920/315276 bytes) Receiving

Download [########-                     ] (90112/315276 bytes) Receiving

Download [#########-                    ] (98304/315276 bytes) Receiving

Download [

In [103]:
# word net import:

# unmark if you want to use:
import wn

### (optional) Hebrew text analysis - hebrew_tokenizer (Tokenizer for Hebrew)
Note: the hebrew_tokenizer is not a must

#### (optional) Only if you didn't install hebrew_tokenizer use:

In [4]:
# Hebrew tokenizer installation:

# unmark if you want to use and need to install:
!pip install hebrew_tokenizer

In [104]:
# Hebrew tokenizer import:

# unmark if you want to use:
import hebrew_tokenizer as ht
from hebrew_tokenizer import groups 

### Reading input files
Reading input files for train annotated corpus (raw text data) corpus and for the test corpus

In [105]:
train_filename = 'annotated_corpus_for_train.csv'
test_filename  = 'corpus_for_test.csv'
df_train = pd.read_csv(train_filename, index_col=None, encoding='utf-8')
df_test  = pd.read_csv(test_filename, index_col=None, encoding='utf-8')

In [106]:
df_train.head(8)
df_train.shape

story gender
0  כשחבר הזמין אותי לחול, לא באמת חשבתי שזה יקרה,...      m
1  לפני שהתגייסתי לצבא עשיתי כל מני מיונים ליחידו...      m
2  מאז שהתחילו הלימודים חלומו של כל סטודנט זה הפנ...      f
3  כשהייתי ילד, מטוסים היה הדבר שהכי ריתק אותי. ב...      m
4  ‏הייתי מדריכה בכפר נוער ומתאם הכפר היינו צריכי...      f
5  לפני כ3 חודשים טסתי לרומא למשך שבוע. טסתי במטו...      f
6  אני כבר שנתיים נשוי והשנה אני ואישתי סוף סוף י...      m
7  השנה התחלנו שיפוץ בדירה שלנו בתל אביב. הדירה ה...      f

(753, 2)

In [107]:
df_test.head(3)
df_test.shape

test_example_id                                              story
0                0  כל קיץ אני והמשפחה נוסעים לארצות הברית לוס אנג...
1                1  הגעתי לשירות המדינה אחרי שנתיים כפעיל בתנועת "...
2                2  אחת האהבות הגדולות שלי אלו הכלבים שלי ושל אישת...

(323, 2)

# Text Analysis Assignment
## Written By: Nadav Bitran Numa


* ***In this notebook we will deal with text analysis. The notebook is divided into the following steps:***
##### - Step 1: The Pre-Processing Phase
##### - Step 2: Vectorization Phase
##### - Step 3: First performance evaluation
##### - Step 4: Perfornamce improvement attempt
##### - Step 5: Final performance evaluation
##### - Step 6: Performing the flow on the test set

## Step 0: Converting the gender column to binary columns

In [108]:
df_train['IsMale'] = df_train['gender'].replace({'m':1,'f':0})
df_train['IsFemale'] = df_train['gender'].replace({'m':0,'f':1})
df_train.drop(columns=['gender'],inplace=True)
df_train

story  IsMale  IsFemale
0    כשחבר הזמין אותי לחול, לא באמת חשבתי שזה יקרה,...       1         0
1    לפני שהתגייסתי לצבא עשיתי כל מני מיונים ליחידו...       1         0
2    מאז שהתחילו הלימודים חלומו של כל סטודנט זה הפנ...       0         1
3    כשהייתי ילד, מטוסים היה הדבר שהכי ריתק אותי. ב...       1         0
4    ‏הייתי מדריכה בכפר נוער ומתאם הכפר היינו צריכי...       0         1
..                                                 ...     ...       ...
748  אז לפני שנה בדיוק טסתי לאמסטרדם עם שני חברים ט...       1         0
749  שבוע שעבר העליתי באופן ספונטני רעיון לנסוע עם ...       1         0
750  לפני חודש עברנו לדירה בבית שמש בעקבות משפחתי ה...       1         0
751  החוויה אותה ארצה לשתף התרחשה לפני כמה חודשים, ...       0         1
752  פעם כשהייתי בחו" "ל ,בקבולומביה כחלק מהטיול של...       1         0

[753 rows x 3 columns]

# 

# 

## Step 1: Pre-Processing

In This Step We'll:

- We will perform **Word Normalization** on the data, in order to leave Hebrew words only, without letters, dates, etc...

- We will perform **Lemmatization** on the data, in order to turn words into their root version

- In addition, we will **remove punctuation** from the corpus

### 1.1: Performing Word Normalization On The Data
* NOTE: We will perform Word Normalization using the Hebrew Tokenizer model

**Before:**

In [109]:
df_train.head(8)

story  IsMale  IsFemale
0  כשחבר הזמין אותי לחול, לא באמת חשבתי שזה יקרה,...       1         0
1  לפני שהתגייסתי לצבא עשיתי כל מני מיונים ליחידו...       1         0
2  מאז שהתחילו הלימודים חלומו של כל סטודנט זה הפנ...       0         1
3  כשהייתי ילד, מטוסים היה הדבר שהכי ריתק אותי. ב...       1         0
4  ‏הייתי מדריכה בכפר נוער ומתאם הכפר היינו צריכי...       0         1
5  לפני כ3 חודשים טסתי לרומא למשך שבוע. טסתי במטו...       0         1
6  אני כבר שנתיים נשוי והשנה אני ואישתי סוף סוף י...       1         0
7  השנה התחלנו שיפוץ בדירה שלנו בתל אביב. הדירה ה...       0         1

In [110]:
def WordNormalize(stories):
    count = 0
    for i in range(len(stories)):
        # We'll Tokinze each story
        tokens_list = ht.tokenize(stories[i])
        fixed_word_list = []
        for grp, token, token_num, (start_index, end_index) in tokens_list:
            # each Tokinzation we'll remove the words that are'nt written on hebrew such as: numbers,dates etc...
            if grp=='HEBREW': 
                count=count+1
                fixed_word_list.append(token)
        stories[i]= ' '.join(fixed_word_list)
    return count

removed = WordNormalize(df_train['story'])

**After:**

In [111]:
df_train.head(8)

story  IsMale  IsFemale
0  כשחבר הזמין אותי לחול לא באמת חשבתי שזה יקרה פ...       1         0
1  לפני שהתגייסתי לצבא עשיתי כל מני מיונים ליחידו...       1         0
2  מאז שהתחילו הלימודים חלומו של כל סטודנט זה הפנ...       0         1
3  כשהייתי ילד מטוסים היה הדבר שהכי ריתק אותי בתו...       1         0
4  הייתי מדריכה בכפר נוער ומתאם הכפר היינו צריכים...       0         1
5  לפני כ חודשים טסתי לרומא למשך שבוע טסתי במטוס ...       0         1
6  אני כבר שנתיים נשוי והשנה אני ואישתי סוף סוף י...       1         0
7  השנה התחלנו שיפוץ בדירה שלנו בתל אביב הדירה הי...       0         1

In [12]:
print("Removed: " , removed , " English Word,Letters,Numbers and so on... From The DS!")

Removed:  250923  English Word,Letters,Numbers and so on... From The DS!


# 

### 1.2: Performing Lemmatization & Removal Of Punctuation From The Data
* NOTE: We'll perform Lemmatization using the Hebrew Tokenizer And WordNet Hebrew Moduls!

In [653]:
def RemovePunctuation(stories):
    punctuations = '!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'
    for i in range(len(punctuations)):
        for j in range(len(stories)):
            stories[j] = stories[j].replace(punctuations[i],"")
    return stories
    
def PreformLemmatization(stories):
    count = 0
    for i in range(len(stories)):
        tokens_list = ht.tokenize(stories[i])
        fixed_word_list = []
        for grp, token, token_num, (start_index, end_index) in tokens_list:
            word = wn.synsets(token)
            if len(word)>0:
                if len(word[0].lemmas()) > 0: # there's more then 1 root to this word
                    fixed_word_list.append(word[0].lemmas()[0])
                else: fixed_word_list.append(word[0].lemmas())
                count=count+1
            else: fixed_word_list.append(token)
        stories[i] = ' '.join(fixed_word_list)
    return count

amount_lemmas = PreformLemmatization(df_train['story'])
WordNormalize(df_train['story'])

251388

# 

## Step 1 Conclusions:
This part is important for the quality of the model itself, if we leave punctuation marks in the corpus, or do not normalize the corpus at all, we will not be able to perform a vectorization that will rely only on the words.
for example:
- In text no. 50, the word will be written: ",אבל"
- In text no. 51, the word will be written: "אבל"

The vectorization we will do, which relies on the Bag Of Words (BOW) model, will interpret them as two different words.

# 

## Step 2: Vectorization 
In this section we will perform **4** different types of vectorization that we can use for comparison later:

- tfIDF Vectorization
- sublinear Vectorization (using log function)
- count Vectorization
- binary Vectorization


### 2.1: Performing tfIDF Vectorization

We will use the following parameters:

- max_features = 2000 -> will limit the number of columns in the dataset to 2000, so that the 2000 with the highest frequency will be chosen

- analyzer= 'word' -> will divide the data by words

- min_df = 3 -> Only the words that appeared in at least 3 texts from the whole corpus will be chosen

- max_df = 0.8 -> will not choose words that appeared in more than 80% of all the texts in the corpus

In [112]:
vectorizer = TfidfVectorizer(use_idf=True,max_features=2000,analyzer='word',min_df=3,max_df=0.8)

data = list(df_train['story'])

df_train_vectorized_data = vectorizer.fit_transform(data)

df_train_vectorized = pd.DataFrame(MinMaxScaler().fit_transform(df_train_vectorized_data.toarray()),columns=vectorizer.get_feature_names_out())


df_train_vectorized['IsMale'] = df_train['IsMale']

df_train_vectorized['IsFemale'] = df_train['IsFemale']


In [113]:
df_train_tfIDF = df_train_vectorized
df_train_tfIDF

אבא   אבי  אביב   אבל  אדום   אדם  אהבתי    או  אוהב  אוהבים  אוהבת  \
0   0.000 0.000 0.000 0.374 0.000 0.000  0.000 0.000 0.000   0.000  0.000   
1   0.000 0.000 0.000 0.105 0.000 0.473  0.000 0.000 0.199   0.000  0.000   
2   0.000 0.000 0.000 0.624 0.000 0.000  0.000 0.000 0.000   0.000  0.000   
3   0.535 0.000 0.000 0.209 0.000 0.156  0.000 0.000 0.000   0.000  0.000   
4   0.000 0.000 0.000 0.237 0.000 0.000  0.000 0.135 0.000   0.000  0.000   
..    ...   ...   ...   ...   ...   ...    ...   ...   ...     ...    ...   
748 0.000 0.000 0.000 0.000 0.000 0.147  0.000 0.112 0.000   0.000  0.000   
749 0.000 0.000 0.000 0.292 0.000 0.000  0.000 0.222 0.000   0.000  0.000   
750 0.000 0.000 0.000 0.244 0.000 0.000  0.000 0.000 0.000   0.000  0.000   
751 0.000 0.000 0.000 0.000 0.000 0.000  0.000 0.108 0.000   0.000  0.000   
752 0.000 0.000 0.000 0.211 0.000 0.000  0.000 0.000 0.000   0.000  0.000   

     אוויר  אווירה  אוטובוס  אוכל  אולי  אולם  אומר  אומרת  אופן  אופנוע  \
0    0.000   0.000    0.000 0.000 0.000 0.000 0.157  0.000 0.000   0.000   
1    0.000   0.000    0.000 0.000 0.000 0.000 0.354  0.000 0.000   0.000   
2    0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
3    0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
4    0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
..     ...     ...      ...   ...   ...   ...   ...    ...   ...     ...   
748  0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
749  0.752   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
750  0.000   0.000    0.000 0.265 0.307 0.000 0.000  0.000 0.000   0.000   
751  0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
752  0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   

     אותה  אותו  אותי  אותך  אותם  אותן  אותנו    אז  אזור    אח   אחד  אחורה  \
0   0.000 0.066 0.111 0.000 0.000 0.000  0.000 0.093 0.000 0.000 0.279  0.000   
1   0.000 0.000 0.000 0.000 0.000 0.000  0.177 0.000 0.000 0.000 0.000  0.000   
2   0.108 0.000 0.000 0.000 0.000 0.000  0.000 0.124 0.000 0.000 0.373  0.000   
3   0.000 0.223 0.249 0.000 0.000 0.000  0.000 0.208 0.000 0.000 0.208  0.306   
4   0.000 0.000 0.704 0.000 0.252 0.000  0.000 0.000 0.000 0.000 0.236  0.000   
..    ...   ...   ...   ...   ...   ...    ...   ...   ...   ...   ...    ...   
748 0.085 0.000 0.000 0.000 0.000 0.000  0.000 0.196 0.000 0.000 0.196  0.000   
749 0.000 0.138 0.000 0.000 0.000 0.000  0.163 0.000 0.000 0.000 0.097  0.000   
750 0.000 0.000 0.145 0.000 0.000 0.000  0.000 0.000 0.000 0.000 0.121  0.000   
751 0.082 0.067 0.000 0.000 0.000 0.000  0.000 0.282 0.000 0.000 0.848  0.000   
752 0.000 0.150 0.000 0.000 0.000 0.000  0.353 0.105 0.000 0.000 0.105  0.000   

     אחותי   אחי  אחלה   אחר  אחראי  אחרות  אחרי  אחריו  אחריות  אחריי  אחרים  \
0    0.000 0.000 0.303 0.000  0.000  0.000 0.103  0.000   0.000  0.000  0.000   
1    0.000 0.000 0.000 0.000  0.000  0.000 0.116  0.000   0.000  0.000  0.000   
2    0.000 0.000 0.000 0.331  0.000  0.000 0.413  0.000   0.000  0.000  0.000   
3    0.000 0.000 0.000 0.000  0.000  0.000 0.115  0.000   0.000  0.000  0.000   
4    0.000 0.000 0.000 0.000  0.000  0.000 0.392  0.000   0.000  0.000  0.000   
..     ...   ...   ...   ...    ...    ...   ...    ...     ...    ...    ...   
748  0.000 0.000 0.000 0.000  0.000  0.000 0.109  0.000   0.000  0.000  0.000   
749  0.000 0.000 0.000 0.000  0.000  0.000 0.107  0.000   0.000  0.000  0.000   
750  0.000 0.000 0.000 0.000  0.000  0.000 0.269  0.000   0.518  0.000  0.000   
751  0.000 0.000 0.000 0.501  0.000  0.000 0.104  0.000   0.000  0.000  0.000   
752  0.000 0.000 0.000 0.000  0.000  0.000 0.000  0.000   0.000  0.000  0.000   

     אחרת   אחת    אי  איזה   איך   אין  איפה  אירוע  אירועים   איש  אישור  \
0   0.000 0.181 0.000 0.000 0.000 0.179 0.300  0.000    0.000 0.000  0.000   
1   0.000 0.000 0.000 0.000 0.000 0.000 0.000  0.000    0.000 0.00

# 

### 2.2: Performing sublinear Vectorization

We will use the following parameters:

- use_idf = False -> It will not perform inverse document frequency, which means that the vectorization will be conducted exactly like the Count Vectorizer which uses term frequency only

- max_features = 2000 -> will limit the number of columns in the dataset to 2000, so that the 2000 with the highest frequency will be chosen

- analyzer= 'word' -> will divide the data by words

- min_df = 3 -> Only the words that appeared in at least 3 texts from the whole corpus will be chosen

- sublinear_tf = True -> will use the following function: log(tf) + 1 instead of counting the tf only. This technique helps us because the larger the text, the more words that appear a few times have more meaning for learning than the words that appear a lot of times

- max_df = 0.8 -> will not choose words that appeared in more than 80% of all the texts in the corpus

In [114]:
vectorizer = TfidfVectorizer(use_idf=False,max_features=2000,analyzer='word',min_df=3,sublinear_tf=True,max_df=0.8)

data = list(df_train['story'])

df_train_vectorized_data = vectorizer.fit_transform(data)

df_train_vectorized = pd.DataFrame(MinMaxScaler().fit_transform(df_train_vectorized_data.toarray()),columns=vectorizer.get_feature_names_out())


df_train_vectorized['IsMale'] = df_train['IsMale']

df_train_vectorized['IsFemale'] = df_train['IsFemale']

In [115]:
df_train_sublinear = df_train_vectorized
df_train_sublinear

אבא   אבי  אביב   אבל  אדום   אדם  אהבתי    או  אוהב  אוהבים  אוהבת  \
0   0.000 0.000 0.000 0.609 0.000 0.000  0.000 0.000 0.000   0.000  0.000   
1   0.000 0.000 0.000 0.291 0.000 0.723  0.000 0.000 0.261   0.000  0.000   
2   0.000 0.000 0.000 0.794 0.000 0.000  0.000 0.000 0.000   0.000  0.000   
3   0.695 0.000 0.000 0.471 0.000 0.329  0.000 0.000 0.000   0.000  0.000   
4   0.000 0.000 0.000 0.529 0.000 0.000  0.000 0.346 0.000   0.000  0.000   
..    ...   ...   ...   ...   ...   ...    ...   ...   ...     ...    ...   
748 0.000 0.000 0.000 0.000 0.000 0.352  0.000 0.329 0.000   0.000  0.000   
749 0.000 0.000 0.000 0.559 0.000 0.000  0.000 0.499 0.000   0.000  0.000   
750 0.000 0.000 0.000 0.533 0.000 0.000  0.000 0.000 0.000   0.000  0.000   
751 0.000 0.000 0.000 0.000 0.000 0.000  0.000 0.286 0.000   0.000  0.000   
752 0.000 0.000 0.000 0.506 0.000 0.000  0.000 0.000 0.000   0.000  0.000   

     אוויר  אווירה  אוטובוס  אוכל  אולי  אולם  אומר  אומרת  אופן  אופנוע  \
0    0.000   0.000    0.000 0.000 0.000 0.000 0.295  0.000 0.000   0.000   
1    0.000   0.000    0.000 0.000 0.000 0.000 0.568  0.000 0.000   0.000   
2    0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
3    0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
4    0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
..     ...     ...      ...   ...   ...   ...   ...    ...   ...     ...   
748  0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
749  0.739   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
750  0.000   0.000    0.000 0.378 0.400 0.000 0.000  0.000 0.000   0.000   
751  0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   
752  0.000   0.000    0.000 0.000 0.000 0.000 0.000  0.000 0.000   0.000   

     אותה  אותו  אותי  אותך  אותם  אותן  אותנו    אז  אזור    אח   אחד  אחורה  \
0   0.000 0.246 0.208 0.000 0.000 0.000  0.000 0.201 0.000 0.000 0.532  0.000   
1   0.000 0.000 0.000 0.000 0.000 0.000  0.270 0.000 0.000 0.000 0.000  0.000   
2   0.281 0.000 0.000 0.000 0.000 0.000  0.000 0.240 0.000 0.000 0.634  0.000   
3   0.000 0.562 0.383 0.000 0.000 0.000  0.000 0.371 0.000 0.000 0.468  0.449   
4   0.000 0.000 0.663 0.000 0.413 0.000  0.000 0.000 0.000 0.000 0.526  0.000   
..    ...   ...   ...   ...   ...   ...    ...   ...   ...   ...   ...    ...   
748 0.274 0.000 0.000 0.000 0.000 0.000  0.000 0.396 0.000 0.000 0.500  0.000   
749 0.000 0.435 0.000 0.000 0.000 0.000  0.248 0.000 0.000 0.000 0.265  0.000   
750 0.000 0.000 0.256 0.000 0.000 0.000  0.000 0.000 0.000 0.000 0.313  0.000   
751 0.239 0.249 0.000 0.000 0.000 0.000  0.000 0.427 0.000 0.000 0.821  0.000   
752 0.000 0.487 0.000 0.000 0.000 0.000  0.470 0.235 0.000 0.000 0.297  0.000   

     אחותי   אחי  אחלה   אחר  אחראי  אחרות  אחרי  אחריו  אחריות  אחריי  אחרים  \
0    0.000 0.000 0.344 0.000  0.000  0.000 0.273  0.000   0.000  0.000  0.000   
1    0.000 0.000 0.000 0.000  0.000  0.000 0.311  0.000   0.000  0.000  0.000   
2    0.000 0.000 0.000 0.458  0.000  0.000 0.683  0.000   0.000  0.000  0.000   
3    0.000 0.000 0.000 0.000  0.000  0.000 0.297  0.000   0.000  0.000  0.000   
4    0.000 0.000 0.000 0.000  0.000  0.000 0.701  0.000   0.000  0.000  0.000   
..     ...   ...   ...   ...    ...    ...   ...    ...     ...    ...    ...   
748  0.000 0.000 0.000 0.000  0.000  0.000 0.318  0.000   0.000  0.000  0.000   
749  0.000 0.000 0.000 0.000  0.000  0.000 0.285  0.000   0.000  0.000  0.000   
750  0.000 0.000 0.000 0.000  0.000  0.000 0.570  0.000   0.607  0.000  0.000   
751  0.000 0.000 0.000 0.659  0.000  0.000 0.276  0.000   0.000  0.000  0.000   
752  0.000 0.000 0.000 0.000  0.000  0.000 0.000  0.000   0.000  0.000  0.000   

     אחרת   אחת    אי  איזה   איך   אין  איפה  אירוע  אירועים   איש  אישור  \
0   0.000 0.335 0.000 0.000 0.000 0.329 0.381  0.000    0.000 0.000  0.000   
1   0.000 0.000 0.000 0.000 0.000 0.000 0.000  0.000    0.000 0.00

# 

### 2.3: Performing count Vectorization

We will use the following parameters:

- max_features = 2000 -> will limit the number of columns in the dataset to 2000, so that the 2000 with the highest frequency will be chosen

- analyzer= 'word' -> will divide the data by words

- min_df = 3 -> Only the words that appeared in at least 3 texts from the whole corpus will be chosen

- max_df = 0.8 -> will not choose words that appeared in more than 80% of all the texts in the corpus

In [116]:
vectorizer = CountVectorizer(max_features=2000,analyzer='word',min_df=3,max_df=0.8)

data = list(df_train['story'])

df_train_vectorized_data = vectorizer.fit_transform(data)

df_train_vectorized = pd.DataFrame(df_train_vectorized_data.toarray(),columns=vectorizer.get_feature_names_out())

df_train_vectorized['IsMale'] = df_train['IsMale']

df_train_vectorized['IsFemale'] = df_train['IsFemale']

In [117]:
df_train_count = df_train_vectorized
df_train_count

אבא  אבי  אביב  אבל  אדום  אדם  אהבתי  או  אוהב  אוהבים  אוהבת  אוויר  \
0      0    0     0    4     0    0      0   0     0       0      0      0   
1      0    0     0    1     0    3      0   0     1       0      0      0   
2      0    0     0    5     0    0      0   0     0       0      0      0   
3      5    0     0    2     0    1      0   0     0       0      0      0   
4      0    0     0    2     0    0      0   1     0       0      0      0   
..   ...  ...   ...  ...   ...  ...    ...  ..   ...     ...    ...    ...   
748    0    0     0    0     0    1      0   1     0       0      0      0   
749    0    0     0    3     0    0      0   2     0       0      0      3   
750    0    0     0    2     0    0      0   0     0       0      0      0   
751    0    0     0    0     0    0      0   1     0       0      0      0   
752    0    0     0    2     0    0      0   0     0       0      0      0   

     אווירה  אוטובוס  אוכל  אולי  אולם  אומר  אומרת  אופן  אופנוע  אותה  אותו  \
0         0        0     0     0     0     1      0     0       0     0     1   
1         0        0     0     0     0     2      0     0       0     0     0   
2         0        0     0     0     0     0      0     0       0     1     0   
3         0        0     0     0     0     0      0     0       0     0     3   
4         0        0     0     0     0     0      0     0       0     0     0   
..      ...      ...   ...   ...   ...   ...    ...   ...     ...   ...   ...   
748       0        0     0     0     0     0      0     0       0     1     0   
749       0        0     0     0     0     0      0     0       0     0     2   
750       0        0     1     1     0     0      0     0       0     0     0   
751       0        0     0     0     0     0      0     0       0     1     1   
752       0        0     0     0     0     0      0     0       0     0     2   

     אותי  אותך  אותם  אותן  אותנו  אז  אזור  אח  אחד  אחורה  אחותי  אחי  \
0       1     0     0     0      0   1     0   0    3      0      0    0   
1       0     0     0     0      1   0     0   0    0      0      0    0   
2       0     0     0     0      0   1     0   0    3      0      0    0   
3       2     0     0     0      0   2     0   0    2      1      0    0   
4       5     0     1     0      0   0     0   0    2      0      0    0   
..    ...   ...   ...   ...    ...  ..   ...  ..  ...    ...    ...  ...   
748     0     0     0     0      0   2     0   0    2      0      0    0   
749     0     0     0     0      1   0     0   0    1      0      0    0   
750     1     0     0     0      0   0     0   0    1      0      0    0   
751     0     0     0     0      0   3     0   0    9      0      0    0   
752     0     0     0     0      2   1     0   0    1      0      0    0   

     אחלה  אחר  אחראי  אחרות  אחרי  אחריו  אחריות  אחריי  אחרים  אחרת  אחת  \
0       1    0      0      0     1      0       0      0      0     0    1   
1       0    0      0      0     1      0       0      0      0     0    0   
2       0    1      0      0     3      0       0      0      0     0    1   
3       0    0      0      0     1      0       0      0      0     0    0   
4       0    0      0      0     3      0       0      0      0     0    2   
..    ...  ...    ...    ...   ...    ...     ...    ...    ...   ...  ...   
748     0    0      0      0     1      0       0      0      0     0    0   
749     0    0      0      0     1      0       0      0      0     0    0   
750     0    0      0      0     2      0       1      0      0     0    0   
751     0    2      0      0     1      0       0      0      0     0    0   
752     0    0      0      0     0      0       0      0      0     0    0   

     אי  איזה  איך  אין  איפה  אירוע  אירועים  איש  אישור  אישי  אישית  אישתי  \
0     0     0    0    1     1      0        0    0      0     0      0      0   
1     0     0    0    0     0      0        0    0      0     0      0      0   
2 

# 

### 2.4: Performing binary Vectorization

We will use the following parameters:

- max_features = 2000 -> will limit the number of columns in the dataset to 2000, so that the 2000 with the highest frequency will be chosen

- analyzer= 'word' -> will divide the data by words

- min_df = 3 -> Only the words that appeared in at least 3 texts from the whole corpus will be chosen

- binary = True -> self explanatory

- max_df = 0.8 -> will not choose words that appeared in more than 80% of all the texts in the corpus

In [118]:
vectorizer = CountVectorizer(max_features=2000,analyzer='word',min_df=3,binary=True,max_df=0.8)

data = list(df_train['story'])

df_train_vectorized_data = vectorizer.fit_transform(data)

df_train_vectorized = pd.DataFrame(df_train_vectorized_data.toarray(),columns=vectorizer.get_feature_names_out())

df_train_vectorized['IsMale'] = df_train['IsMale']

df_train_vectorized['IsFemale'] = df_train['IsFemale']

In [119]:
df_train_binary = df_train_vectorized
df_train_binary

אבא  אביב  אבל  אדום  אדם  אהבתי  או  אוהב  אוהבים  אוהבת  אוויר  אווירה  \
0      0     0    1     0    0      0   0     0       0      0      0       0   
1      0     0    1     0    1      0   0     1       0      0      0       0   
2      0     0    1     0    0      0   0     0       0      0      0       0   
3      1     0    1     0    1      0   0     0       0      0      0       0   
4      0     0    1     0    0      0   1     0       0      0      0       0   
..   ...   ...  ...   ...  ...    ...  ..   ...     ...    ...    ...     ...   
748    0     0    0     0    1      0   1     0       0      0      0       0   
749    0     0    1     0    0      0   1     0       0      0      1       0   
750    0     0    1     0    0      0   0     0       0      0      0       0   
751    0     0    0     0    0      0   1     0       0      0      0       0   
752    0     0    1     0    0      0   0     0       0      0      0       0   

     אוטובוס  אוכל  אולי  אולם  אומנם  אומר  אומרת  אופן  אותה  אותו  אותי  \
0          0     0     0     0      1     1      0     0     0     1     1   
1          0     0     0     0      0     1      0     0     0     0     0   
2          0     0     0     0      0     0      0     0     1     0     0   
3          0     0     0     0      0     0      0     0     0     1     1   
4          0     0     0     0      0     0      0     0     0     0     1   
..       ...   ...   ...   ...    ...   ...    ...   ...   ...   ...   ...   
748        0     0     0     0      0     0      0     0     1     0     0   
749        0     0     0     0      1     0      0     0     0     1     0   
750        0     1     1     0      1     0      0     0     0     0     1   
751        0     0     0     0      0     0      0     0     1     1     0   
752        0     0     0     0      0     0      0     0     0     1     0   

     אותך  אותם  אותן  אותנו  אז  אזור  אח  אחד  אחורה  אחותי  אחזור  אחי  \
0       0     0     0      0   1     0   0    1      0      0      0    0   
1       0     0     0      1   0     0   0    0      0      0      0    0   
2       0     0     0      0   1     0   0    1      0      0      0    0   
3       0     0     0      0   1     0   0    1      1      0      0    0   
4       0     1     0      0   0     0   0    1      0      0      1    0   
..    ...   ...   ...    ...  ..   ...  ..  ...    ...    ...    ...  ...   
748     0     0     0      0   1     0   0    1      0      0      0    0   
749     0     0     0      1   0     0   0    1      0      0      0    0   
750     0     0     0      0   0     0   0    1      0      0      0    0   
751     0     0     0      0   1     0   0    1      0      0      0    0   
752     0     0     0      1   1     0   0    1      0      0      0    0   

     אחלה  אחר  אחראי  אחרות  אחרי  אחריו  אחריות  אחרים  אחרת  אחת  אי  איזה  \
0       1    0      0      0     1      0       0      0     0    1   0     0   
1       0    0      0      0     1      0       0      0     0    0   0     0   
2       0    1      0      0     1      0       0      0     0    1   0     0   
3       0    0      0      0     1      0       0      0     0    0   1     1   
4       0    0      0      0     1      0       0      0     0    1   0     0   
..    ...  ...    ...    ...   ...    ...     ...    ...   ...  ...  ..   ...   
748     0    0      0      0     1      0       0      0     0    0   0     0   
749     0    0      0      0     1      0       0      0     0    0   0     0   
750     0    0      0      0     1      0       1      0     0    0   0     1   
751     0    1      0      0     1      0       0      0     0    0   0     0   
752     0    0      0      0     0      0       0      0     0    0   0     1   

     איך  אין  איפה  אירוע  איש  אישור  אישי  אישית  אישתי  איתה  איתו  איתי  \
0      0    1     1      0    0      0     0      0      0     0     1     0   
1      0    0     0      0    0      

# 

## Step 2 Conclusions:
The creation of these four vectorizations will help us further to look at the different machine learning algorithms while understanding which algorithms are best for the different types of vectorizations

**Important note**:
In count and binary we did not perform scaling for the following reasons:
* For count vectorizer we will stay with discrete variables to see how machine learning algorithms (such as multinomialNB) will produce results.
* For binary vectorizer there is no need to perform scaling

# 

## Step 3: Preliminary comparisons between types of machine learning algorithems (MODEL SELECTION)
In this section we will make comparisons between different machine learning algorithms, below are the algorithms we will use:
- KNN
- Decision Tree
- GaussianNB
- LinearSVC
- MLPClassifier
- Perception
- SGDClassifier
- MultinomialNB

In this part, we will go through all the algorithms individually, while performing a performance evaluation with the help of Validation Set by Cross Validation.
The performance evaluation will be performed on all types of datasets that we prepared in part 2 (tfIDF, count, binary, sublinear).
Finally, we will look at all the performance evaluations we received and draw conclusions

**NOTE: In this part, the performance evaluation will be performed by different algorithems with default hyperparameters, later we will perform tuning**



**But first of all, we will write a function that we will use in this part, which will perform 10-FoldCrossValidation for us**

### 3.0.1: 10-FoldCrossValidation Function

**Input**: The database and algorithm for which we will perform the performance evaluation.

**Output**: a measure for evaluating performance

* We will remove the columns 'IsMale,' and 'IsFemale' from the database
* We will perform a 10Fold performance evaluation separately for each gender.
* We will get f1 indices for male and female
* We will return the **macro average f1** which is an average of the f1 indices of male and female

In [120]:
def PreformTenFoldCrossValidation(dataset,MLalgo):
    X = dataset[dataset.columns[(dataset.columns!='IsMale') & (dataset.columns!='IsFemale')]]
    y_male = dataset['IsMale']
    y_female = dataset['IsFemale']
    f1_male = cross_val_score(MLalgo,X,y_male,cv=10,scoring='f1').mean()
    f1_female = cross_val_score(MLalgo,X,y_female,cv=10,scoring='f1').mean()
    return (f1_male + f1_female) / 2

# 

### 3.0.2: Creating a 'scores' matrix

In [121]:
types = ['Binary','Count','Sublinear','TfIDF']
MLalgos_titles = ['KNN','DecisionTree','GussianNB','MultinomialNB','LinearSVC','MLPClassifier','Perceptron','SGDClassifier']
scores = pd.DataFrame(index=types,columns=MLalgos_titles)

# 

### 3.1: KNN


#### 3.1.1: KNN: CrossVal On Binary Vectorization DataSet

In [122]:
scores.loc['Binary','KNN'] = PreformTenFoldCrossValidation(df_train_binary,KNeighborsClassifier())

#### 3.1.2: KNN: CrossVal On Count Vectorization DataSet

In [123]:
scores.loc['Count','KNN'] = PreformTenFoldCrossValidation(df_train_count,KNeighborsClassifier())

#### 3.1.3: KNN: CrossVal On Sublinear Vectorization DataSet

In [124]:
scores.loc['Sublinear','KNN'] = PreformTenFoldCrossValidation(df_train_sublinear,KNeighborsClassifier())

#### 3.1.4: KNN: CrossVal On Tf*IDF Vectorization DataSet

In [125]:
scores.loc['TfIDF','KNN'] = PreformTenFoldCrossValidation(df_train_tfIDF,KNeighborsClassifier())

# 

#### 3.1.5: Results

In [126]:
scores['KNN'].sort_values(ascending=False)

Count       0.533
Sublinear   0.519
Binary      0.514
TfIDF       0.443
Name: KNN, dtype: object

### KNN Conclusions:

* Since the KNN algorithm is a **lazy algorithm**, which loads the entire database into memory for the learning process itself, it is inefficient for this type of database that contains many columns


* The KNN algorithm in general did not produce good results
* **Reason**: It may be that when we perform HyperParameter Tuning it will return better results, and if even then we can't conclude that it is not good for our problem

# 

### 3.2: Decision Tree

#### 3.2.1:  Decision Tree: CrossVal On Binary Vectorization DataSet

In [127]:
scores.loc['Binary','DecisionTree'] = PreformTenFoldCrossValidation(df_train_binary,DecisionTreeClassifier())

#### 3.2.2:  Decision Tree: CrossVal On Count Vectorization DataSet

In [128]:
scores.loc['Count','DecisionTree'] = PreformTenFoldCrossValidation(df_train_count,DecisionTreeClassifier())

#### 3.2.3:  Decision Tree: CrossVal On Sublinear Vectorization DataSet

In [129]:
scores.loc['Sublinear','DecisionTree'] = PreformTenFoldCrossValidation(df_train_sublinear,DecisionTreeClassifier())

#### 3.2.4:  Decision Tree: CrossVal On Tf*IDF Vectorization DataSet

In [130]:
scores.loc['TfIDF','DecisionTree'] = PreformTenFoldCrossValidation(df_train_tfIDF,DecisionTreeClassifier())

# 

#### 3.1.5: Results

In [131]:
scores['DecisionTree'].sort_values(ascending=False)

Count       0.612
Binary      0.590
TfIDF       0.557
Sublinear   0.556
Name: DecisionTree, dtype: object

### Decision Tree Conclusions:

* You can see that the results for Decision Tree are better than KNN, but still not high enough. 
* **Possible Cause**: DataSet size and number of columns affect the quality of the decision tree you can create

# 

### 3.3: GaussianNB

#### 3.3.1: GaussianNB: CrossVal On Binary Vectorization DataSet

In [132]:
scores.loc['Binary','GussianNB'] = PreformTenFoldCrossValidation(df_train_binary,GaussianNB())

#### 3.3.2: GaussianNB: CrossVal On Count Vectorization DataSet

In [133]:
scores.loc['Count','GussianNB'] = PreformTenFoldCrossValidation(df_train_count,GaussianNB())

#### 3.3.3: GaussianNB: CrossVal On Sublinear Vectorization DataSet

In [134]:
scores.loc['Sublinear','GussianNB'] = PreformTenFoldCrossValidation(df_train_sublinear,GaussianNB())

#### 3.3.4: GaussianNB: CrossVal On Tf*IDF Vectorization DataSet

In [135]:
scores.loc['TfIDF','GussianNB'] = PreformTenFoldCrossValidation(df_train_tfIDF,GaussianNB())

# 

#### 3.3.5: Results

In [136]:
scores['GussianNB'].sort_values(ascending=False)

TfIDF       0.649
Count       0.639
Sublinear   0.637
Binary      0.621
Name: GussianNB, dtype: object

### GaussianNB Conclusions:

* It can be seen that the results for GaussianNB are much better than all the algorithms we have tested so far.
* **Possible reason**: the columns are almost independent of each other, which allows Naive Bayes to work well.

# 

### 3.4: MultinomialNB

#### 3.4.1: MultinomialNB: CrossVal On Binary Vectorization DataSet

In [137]:
scores.loc['Binary','MultinomialNB'] = PreformTenFoldCrossValidation(df_train_binary,MultinomialNB())

#### 3.4.2: MultinomialNB: CrossVal On Count Vectorization DataSet

In [138]:
scores.loc['Count','MultinomialNB'] = PreformTenFoldCrossValidation(df_train_count,MultinomialNB())

#### 3.4.3: MultinomialNB: CrossVal On Sublinear Vectorization DataSet

In [139]:
scores.loc['Sublinear','MultinomialNB'] = PreformTenFoldCrossValidation(df_train_sublinear,MultinomialNB())

#### 3.4.4: MultinomialNB: CrossVal On Tf*IDF Vectorization DataSet

In [140]:
scores.loc['TfIDF','MultinomialNB'] = PreformTenFoldCrossValidation(df_train_tfIDF,MultinomialNB())

# 

#### 3.4.5: Results

In [141]:
scores['MultinomialNB'].sort_values(ascending=False)

Count       0.690
Sublinear   0.667
Binary      0.666
TfIDF       0.663
Name: MultinomialNB, dtype: object

### MultinomialNB Conclusions:
Since this algorithm produces particularly good results for text analysis with discontinuous data, it is understandable why we get about 70 percent.

# 

### 3.5: Extra - Additional Algorithms

#### 3.5.1: Additional Algorithms: CrossVal On Binary Vectorization DataSet

In [142]:
scores.loc['Binary','LinearSVC'] = PreformTenFoldCrossValidation(df_train_binary,LinearSVC())
scores.loc['Binary','MLPClassifier'] = PreformTenFoldCrossValidation(df_train_binary,MLPClassifier())
scores.loc['Binary','Perceptron'] = PreformTenFoldCrossValidation(df_train_binary,Perceptron())
scores.loc['Binary','SGDClassifier'] = PreformTenFoldCrossValidation(df_train_binary,SGDClassifier())

#### 3.5.2: Additional Algorithms: CrossVal On Count Vectorization DataSet


In [143]:
scores.loc['Count','LinearSVC'] = PreformTenFoldCrossValidation(df_train_count,LinearSVC())
scores.loc['Count','MLPClassifier'] = PreformTenFoldCrossValidation(df_train_count,MLPClassifier())
scores.loc['Count','Perceptron'] = PreformTenFoldCrossValidation(df_train_count,Perceptron())
scores.loc['Count','SGDClassifier'] = PreformTenFoldCrossValidation(df_train_count,SGDClassifier())

#### 3.5.3: Additional Algorithms: CrossVal On Sublinear Vectorization DataSet

In [144]:
scores.loc['Sublinear','LinearSVC'] = PreformTenFoldCrossValidation(df_train_sublinear,LinearSVC())
scores.loc['Sublinear','MLPClassifier'] = PreformTenFoldCrossValidation(df_train_sublinear,MLPClassifier())
scores.loc['Sublinear','Perceptron'] = PreformTenFoldCrossValidation(df_train_sublinear,Perceptron())
scores.loc['Sublinear','SGDClassifier'] = PreformTenFoldCrossValidation(df_train_sublinear,SGDClassifier())

#### 3.5.4: Additional Algorithms: CrossVal On Tf*IDF Vectorization DataSet

In [145]:
scores.loc['TfIDF','LinearSVC'] = PreformTenFoldCrossValidation(df_train_tfIDF,LinearSVC())
scores.loc['TfIDF','MLPClassifier'] = PreformTenFoldCrossValidation(df_train_tfIDF,MLPClassifier())
scores.loc['TfIDF','Perceptron'] = PreformTenFoldCrossValidation(df_train_tfIDF,Perceptron())
scores.loc['TfIDF','SGDClassifier'] = PreformTenFoldCrossValidation(df_train_tfIDF,SGDClassifier())

# 

#### 3.5.5: Results

In [146]:
scores['LinearSVC'].sort_values(ascending=False)

Sublinear   0.704
Binary      0.697
TfIDF       0.669
Count       0.660
Name: LinearSVC, dtype: object

In [147]:
scores['MLPClassifier'].sort_values(ascending=False)

Sublinear   0.661
Count       0.654
TfIDF       0.647
Binary      0.641
Name: MLPClassifier, dtype: object

In [148]:
scores['Perceptron'].sort_values(ascending=False)

Binary      0.679
Count       0.643
Sublinear   0.642
TfIDF       0.633
Name: Perceptron, dtype: object

In [149]:
scores['SGDClassifier'].sort_values(ascending=False)

Sublinear   0.673
Count       0.651
TfIDF       0.645
Binary      0.643
Name: SGDClassifier, dtype: object

# 

## Preliminary comparisons between types of machine learning algorithems (MODEL SELECTION) - Conclusions
After going through each algorithm separately and drawing conclusions: now we will look at the full picture, look at all the results in one matrix and understand the following:
* Which algorithms and which DataSets produced the best results? and why?
* Which algorithms and which DataSets produced the worst results? and why?
* Finally, based on the results we will try to reach conclusions and choose the four best combinations that will move with us to the next stage - the HyperParameter Tuning stage

----------------------------------------------------------
 
Top combinations that produced the best results:

**1) LinearSVC (df_train_sublinear)     ---> 70.4%**

**2) LinearSVC (df_train_binary)        ---> 69.6%**

**3) MultinomialNB (df_train_count)     ---> 69.0%**

**4) Perceptron (df_train_binary)       ---> 67.9%**

**5) SGDClassifier (df_train_binary) ---> 67.5%**

**6) GaussianNB (df_train_tfIDF)        ---> 65.0%**


----------------------------------------------------------


Worst 3 combinations that producted the worst results:

1) MultinomialNB (df_train_tfIDF) ---> 43.9%

**Explanation why**: This algorithm works well with discrete variables only

2) KNN (df_train_binary) ---> 51.4%

**Explanation why**: KNN algorithm works with vector distances, it may not give good results for a database with only 1's and 0's

3) DecisionTree (df_train_tfIDF) ---> 56.1%

**Explanation why**: decision trees are **discriminative algorithms**, meaning that they look for a separation between the values that will yield a good prediction.
It is possible that the combination of the search for this separation with the fact that the data after using IDF is very dense is simply a bad combination

----------------------------------------------------------

In [150]:
scores

KNN DecisionTree GussianNB MultinomialNB LinearSVC MLPClassifier  \
Binary    0.514        0.590     0.621         0.666     0.697         0.641   
Count     0.533        0.612     0.639         0.690     0.660         0.654   
Sublinear 0.519        0.556     0.637         0.667     0.704         0.661   
TfIDF     0.443        0.557     0.649         0.663     0.669         0.647   

          Perceptron SGDClassifier  
Binary         0.679         0.643  
Count          0.643         0.651  
Sublinear      0.642         0.673  
TfIDF          0.633         0.645

# 

## Step 4: Perfornamce improvement
In this part we will try to improve the performance of the various algorithms.

**Our goal is:**
* Improve the performance of as many algorithms as possible
* To reach over 70% in some of the performance of the algorithms in the maro f1 average 

**How do we'll do it?:**
* Using Grid Search to find the best hyperparameters for each algorithm
* With the help of certain techniques that should lead to learning itself for certain algorithms


Before we start, you can see that for each learning algorithm in the previous step we performed 4 different performance evaluations.
Which means that at this stage we will have to do 32 GridSearchs!!, this is very inefficient and therefore we would like to transfer only some of the algorithms to this stage, which will leads us to the next question:

**What combinations of algorithms with DataSets will we transfer to this stage?**

* For each algorithm we selected the DataSet for which the performance evaluation was the highest. (or several of them)

* We will perform a GridSearch on everyone who is selected

* We will perform additional tests in the hope that they will improve the performance

* We will compare the new results with the old results to see how much we managed to improve the quality of the model

* **NOTE**: For the algorithms we learned in class, we will select all the options

Here are the algorithms chosen to proceed to this stage:
- LinearSVC using ALL
- MultinomialNB using df_train_count
- Perceptron using df_train_binary
- SGDClassifier using ALL
- GaussianNB using ALL
- DecisionTree using ALL
- KNN using ALL

### 4.1: Last improvements

#### 4.1.1: Feature Selection On GaussianNB
In this part our goal is to remove as many words as possible from the database with high similarity to each other, in order to improve the performance of the algorithm.
How do we do it? Using **Pearson Correlation**

The method we will use:
When we found 2 words with high correlation, we will leave only one of them

**NOTE**: Pearson's correlation coefficient examines a linear relationship between 2 **continuous variables** only, so we will perform this feature selection on tfIDf and sublinear only

Explanation of the function:
We will create a matrix with the correlation coefficient values.
If we found 2 words with a high correlation (over 0.7) we will remove the one that has a lower impact in relation to the target label

In [151]:
def getFeaturesToRemove(chosen_df):
    features_to_remove = []
    corr_df = chosen_df.corr()
    rows = list(corr_df.index.values[(corr_df.index.values!='IsMale') & (corr_df.index.values!='IsFemale')])
    cols = list(corr_df.columns.values[(corr_df.columns.values!='IsMale') & (corr_df.columns.values!='IsFemale')])
    for feature_row in rows:
        for feature_col in cols:
            if (corr_df.loc[feature_row,feature_col] >= 0.7) & (feature_row!=feature_col):
                if abs(corr_df.loc[feature_row,'IsMale']) > abs(corr_df.loc[feature_col,'IsMale']):
                    features_to_remove.append(feature_col)
                else: features_to_remove.append(feature_row)
    return features_to_remove

df_train_tfIDF_gnb = df_train_tfIDF.copy()
df_train_sublinear_gnb = df_train_sublinear.copy()

features_to_remove = list(set(getFeaturesToRemove(df_train_tfIDF_gnb)))
df_train_tfIDF_gnb.drop(columns=features_to_remove,inplace=True)

features_to_remove = list(set(getFeaturesToRemove(df_train_sublinear_gnb)))
df_train_sublinear_gnb.drop(columns=features_to_remove,inplace=True)

#### 4.1.2: Feature Selection On GaussianNB - Part 2
In this part, our goal is to remove as many documents as possible that are highly similar to each other, in order to improve the performance of the algorithm.
How do we do it? With the help of **Cosine Similarity**

Explanation of the function: we will create a matrix with the cosine similarity values. If we found two documents with a high similarity to each other (more than 0.7) whose Target Label is the same, we will remove one of them.

In [152]:
def getDocsToRemove(chosen_df):
    docs_to_remove = []
    sim_df = pd.DataFrame(metrics.pairwise.cosine_similarity(chosen_df),columns=chosen_df.index,index=chosen_df.index)
    rows = list(chosen_df.index)
    cols = list(chosen_df.index)
    for doc_row in rows:
        for doc_col in cols:
            if((sim_df.loc[doc_row,doc_col] >= 0.7) & (doc_row!=doc_col) & (chosen_df.loc[doc_row,'IsMale'] == 
                                                                           chosen_df.loc[doc_col,'IsMale'])):
                docs_to_remove.append(doc_row)
    return docs_to_remove

df_train_binary_gnb = df_train_binary.copy()
df_train_count_gnb = df_train_count.copy()

docs_to_remove = list(set(getDocsToRemove(df_train_binary_gnb)))
df_train_binary_gnb.drop(index=docs_to_remove,inplace=True)

docs_to_remove = list(set(getDocsToRemove(df_train_count_gnb)))
df_train_count_gnb.drop(index=docs_to_remove,inplace=True)

docs_to_remove = list(set(getDocsToRemove(df_train_sublinear_gnb)))
df_train_sublinear_gnb.drop(index=docs_to_remove,inplace=True)

docs_to_remove = list(set(getDocsToRemove(df_train_tfIDF_gnb)))
df_train_tfIDF_gnb.drop(index=docs_to_remove,inplace=True)

### 4.2: Grid-Search 

In this section we will perform a GridSearch on some of the algorithms.

**Note:** Running GridSearch on all the data took a very long time.

**But first of all, we will create a function that we will work with in this part that will perform the GridSearch for us.**

#### 4.2.0.1: Grid-Search Function

**Input**: the dataset, machine learning algorithm, hyperparameters dictionary.

**Output**: the best hyper-parameters for this algorithm and dataset.

* Will perform a GridSearch on 10-FoldCrossValidation.
* Will find the best combination of hyper-parameters according to the macro average f1 index
* Will return these parameters

In [175]:
def PreformGridSearch(dataset,algo,dict_params):
    X = dataset[dataset.columns[(dataset.columns!='IsMale') & (dataset.columns!='IsFemale')]]
    y = dataset['IsMale']
    gsCV = GridSearchCV(algo,dict_params,cv=10,scoring='f1_macro')
    gsCVresult = gsCV.fit(X,y)
    return gsCVresult.best_params_

#### 4.2.1: KNN


#### 4.2.1.1: Selected Parameters

In [176]:
dictionary_knn_params = {'n_neighbors' : list(range(3,31,2)),
                         'weights' : ['voting','uniform'],
                         'metric' : ['euclidean', 'manhattan']}

#### 4.2.1.2: Grid-Search

In [177]:
best_params_binary = PreformGridSearch(df_train_binary,KNeighborsClassifier(),dictionary_knn_params)
best_params_count = PreformGridSearch(df_train_count,KNeighborsClassifier(),dictionary_knn_params)
best_params_sublinear = PreformGridSearch(df_train_sublinear,KNeighborsClassifier(),dictionary_knn_params)
best_params_tfIDF = PreformGridSearch(df_train_tfIDF,KNeighborsClassifier(),dictionary_knn_params)

In [178]:
print("BINARY: " , best_params_binary)
print("COUNT: ", best_params_count)
print("SUBLINEAR: ", best_params_sublinear)
print("TFIDF: ", best_params_tfIDF)

BINARY:  {'metric': 'euclidean', 'n_neighbors': 5, 'weights': 'uniform'}
COUNT:  {'metric': 'euclidean', 'n_neighbors': 3, 'weights': 'uniform'}
SUBLINEAR:  {'metric': 'euclidean', 'n_neighbors': 9, 'weights': 'uniform'}
TFIDF:  {'metric': 'manhattan', 'n_neighbors': 5, 'weights': 'uniform'}


# 

#### 4.2.2: DecisionTree

#### 4.2.2.1: Selected Parameters

In [179]:
dictionary_decisionTree_params = {'max_depth' : list(range(1,11)),
                                 'min_samples_split' : list(range(2,11)),
                                 'min_samples_leaf' : list(range(2,6))}

#### 4.2.2.2: Grid-Search

In [180]:
best_params_binary = PreformGridSearch(df_train_binary,DecisionTreeClassifier(),dictionary_decisionTree_params)
best_params_count = PreformGridSearch(df_train_count,DecisionTreeClassifier(),dictionary_decisionTree_params)
best_params_sublinear = PreformGridSearch(df_train_sublinear,DecisionTreeClassifier(),dictionary_decisionTree_params)
best_params_tfIDF = PreformGridSearch(df_train_tfIDF,DecisionTreeClassifier(),dictionary_decisionTree_params)

In [181]:
print("BINARY: " , best_params_binary)
print("COUNT: ", best_params_count)
print("SUBLINEAR: ", best_params_sublinear)
print("TFIDF: ", best_params_tfIDF)

BINARY:  {'max_depth': 10, 'min_samples_leaf': 5, 'min_samples_split': 6}
COUNT:  {'max_depth': 7, 'min_samples_leaf': 4, 'min_samples_split': 4}
SUBLINEAR:  {'max_depth': 8, 'min_samples_leaf': 3, 'min_samples_split': 9}
TFIDF:  {'max_depth': 9, 'min_samples_leaf': 2, 'min_samples_split': 6}


# 

#### 4.2.3: MultinomialNB

#### 4.2.3.1: Selected Parameters

In [182]:
dictionary_multinomialnb_params = {
    'alpha': [0.1, 0.5, 1.0, 2.0, 5.0],
    'fit_prior': [True, False],
}

#### 4.2.3.2: Grid-Search

In [183]:
best_params_count = PreformGridSearch(df_train_count,MultinomialNB(),dictionary_multinomialnb_params)

In [184]:
print("COUNT: ", best_params_count)

COUNT:  {'alpha': 1.0, 'fit_prior': True}


# 

#### 4.2.4: LinearSVC

#### 4.2.4.1: Selected Parameters

In [185]:
dictionary_linearsvc_params = {
    'C': [0.1, 0.5, 1.0, 2.0, 5.0],
    'loss': ['hinge', 'squared_hinge'],
    'max_iter': [500, 1000, 2000],
}

#### 4.2.4.2: Grid-Search

In [186]:
best_params_binary = PreformGridSearch(df_train_binary,LinearSVC(),dictionary_linearsvc_params)
best_params_count = PreformGridSearch(df_train_count,LinearSVC(),dictionary_linearsvc_params)
best_params_sublinear = PreformGridSearch(df_train_sublinear,LinearSVC(),dictionary_linearsvc_params)
best_params_tfIDF = PreformGridSearch(df_train_tfIDF,LinearSVC(),dictionary_linearsvc_params)

In [187]:
print("BINARY: " , best_params_binary)
print("COUNT: ", best_params_count)
print("SUBLINEAR: " , best_params_binary)
print("TFIDF: ", best_params_count)

BINARY:  {'C': 0.5, 'loss': 'hinge', 'max_iter': 500}
COUNT:  {'C': 0.1, 'loss': 'hinge', 'max_iter': 500}
SUBLINEAR:  {'C': 0.5, 'loss': 'hinge', 'max_iter': 500}
TFIDF:  {'C': 0.1, 'loss': 'hinge', 'max_iter': 500}


# 

#### 4.2.5: Perceptron

#### 4.2.5.1: Selected Parameters

In [188]:
dictionary_perceptron_params = {
    'penalty': [None, 'l2', 'l1', 'elasticnet'],
    'alpha': [0.0001, 0.001, 0.01],
    'max_iter': [500, 1000, 2000],
}

#### 4.2.5.2: Grid-Search

In [189]:
best_params_binary = PreformGridSearch(df_train_binary,Perceptron(),dictionary_perceptron_params)

In [190]:
print("BINARY: " , best_params_binary)

BINARY:  {'alpha': 0.0001, 'max_iter': 500, 'penalty': None}


# 

#### 4.2.6: SGD Classifier

#### 4.2.6.1: Selected Parameters

In [191]:
dictionary_sgdc_params = {
    'loss': ['hinge', 'log', 'modified_huber'],
    'penalty': [None, 'l2', 'l1', 'elasticnet'],
    'alpha': [0.0001, 0.001, 0.01],
    'max_iter': [500, 1000, 2000],
}

#### 4.2.6.2: Grid-Search

In [192]:
best_params_binary = PreformGridSearch(df_train_binary,SGDClassifier(),dictionary_sgdc_params)
best_params_count = PreformGridSearch(df_train_count,SGDClassifier(),dictionary_sgdc_params)
best_params_sublinear = PreformGridSearch(df_train_sublinear,SGDClassifier(),dictionary_sgdc_params)
best_params_tfIDF = PreformGridSearch(df_train_tfIDF,SGDClassifier(),dictionary_sgdc_params)

In [193]:
print("BINARY: " , best_params_binary)
print("COUNT: ", best_params_count)
print("SUBLINEAR: ", best_params_sublinear)
print("TFIDF: ", best_params_tfIDF)

BINARY:  {'alpha': 0.01, 'loss': 'hinge', 'max_iter': 500, 'penalty': 'l2'}
COUNT:  {'alpha': 0.0001, 'loss': 'modified_huber', 'max_iter': 500, 'penalty': 'elasticnet'}
SUBLINEAR:  {'alpha': 0.001, 'loss': 'log', 'max_iter': 500, 'penalty': None}
TFIDF:  {'alpha': 0.0001, 'loss': 'hinge', 'max_iter': 2000, 'penalty': 'l1'}


# 

## Step 4 Conclusions
In this section we performed a GridSearch for all the algorithms selected to be used for the final performance evaluations in the next section.
* **Note**: Since the normal Naive Bayes algorithm has no hyper-parameters, we performed other treatments on them to try to improve their performance

# 

## Step 5: Final performance evaluation
In this part we will look at the final performance evaluation for the datasets algorithms selected so that we can improve them in step 4.
* We will look at all the results and make a comparison
* We will look at the top 5 best results
* We will look at the differences between the results after the hyper-parameter tuning and those before
and finally, we'll draw conclusions

### 5.1: Performance Evaluation Using CrossVal

#### 5.1.1: Creating ScoreV2 Matrix
**Note**: The matrix will only contain data on the selected algorithms and datasets, otherwise NaN will be marked

In [194]:
types = ['Binary','Count','Sublinear','TfIDF']
MLalgos_titles = ['KNN','DecisionTree','GussianNB','MultinomialNB','LinearSVC','MLPClassifier','Perceptron','SGDClassifier']
scores_v2 = pd.DataFrame(index=types,columns=MLalgos_titles)

# 

#### 5.1.2: KNN

In [195]:
binary_knn_params = {'metric': 'euclidean', 'n_neighbors': 5, 'weights': 'uniform'}
count_knn_params =  {'metric': 'euclidean', 'n_neighbors': 3, 'weights': 'uniform'}
sublinear_knn_params =  {'metric': 'euclidean', 'n_neighbors': 9, 'weights': 'uniform'}
tfidf_knn_params =  {'metric': 'manhattan', 'n_neighbors': 5, 'weights': 'uniform'}

diff_matrix = pd.DataFrame(columns=['Before','Now'],index=['Binary','Count','Sublinear','TfIDF'])

scores_v2.loc['Binary','KNN'] = PreformTenFoldCrossValidation(df_train_binary,KNeighborsClassifier(
    metric='euclidean',
    n_neighbors=5,
    weights = 'uniform'))

scores_v2.loc['Count','KNN'] = PreformTenFoldCrossValidation(df_train_count,KNeighborsClassifier(
    metric='euclidean',
    n_neighbors=3,
    weights = 'uniform'))

scores_v2.loc['Sublinear','KNN'] = PreformTenFoldCrossValidation(df_train_sublinear,KNeighborsClassifier(
    metric='euclidean',
    n_neighbors=9,
    weights = 'uniform'))

scores_v2.loc['TfIDF','KNN'] = PreformTenFoldCrossValidation(df_train_tfIDF,KNeighborsClassifier(
    metric='manhattan',
    n_neighbors=5,
    weights = 'uniform'))


diff_matrix['Before'] = scores['KNN']
diff_matrix['Now'] = scores_v2['KNN']

In [196]:
diff_matrix

Before   Now
Binary     0.514 0.514
Count      0.533 0.539
Sublinear  0.519 0.595
TfIDF      0.443 0.512

You can see that the differences for tfidf and sublinear are very large and reach up to about 8 percent!
But for count and binary there is no difference at all because KNN is simply not suitable for them

In [197]:
print("DIFFERENCES BY PRECENT:")
((diff_matrix['Now'] - diff_matrix['Before'])*100).sort_values(ascending=False)

DIFFERENCES BY PRECENT:


Sublinear   7.581
TfIDF       6.880
Count       0.592
Binary      0.000
dtype: object

# 

#### 5.2.2: DecisionTree

In [198]:
binary_params =  {'max_depth': 10, 'min_samples_leaf': 5, 'min_samples_split': 6}
count_params = {'max_depth': 7, 'min_samples_leaf': 4, 'min_samples_split': 4}
sublinear_params = {'max_depth': 8, 'min_samples_leaf': 3, 'min_samples_split': 9}
tfidf_params = {'max_depth': 9, 'min_samples_leaf': 2, 'min_samples_split': 6}


scores_v2.loc['Binary','DecisionTree'] = PreformTenFoldCrossValidation(df_train_binary,DecisionTreeClassifier(
    max_depth=10,
    min_samples_leaf=5,
    min_samples_split =6))

scores_v2.loc['Count','DecisionTree'] = PreformTenFoldCrossValidation(df_train_count,DecisionTreeClassifier(
    max_depth=7,
    min_samples_leaf=4,
    min_samples_split =4))

scores_v2.loc['Sublinear','DecisionTree'] = PreformTenFoldCrossValidation(df_train_sublinear,DecisionTreeClassifier(
    max_depth=8,
    min_samples_leaf=3,
    min_samples_split = 9))

scores_v2.loc['TfIDF','DecisionTree'] = PreformTenFoldCrossValidation(df_train_tfIDF,DecisionTreeClassifier(
    max_depth=9,
    min_samples_leaf=2,
    min_samples_split = 6))


diff_matrix['Before'] = scores['DecisionTree']
diff_matrix['Now'] = scores_v2['DecisionTree']

In [199]:
diff_matrix

Before   Now
Binary     0.590 0.592
Count      0.612 0.577
Sublinear  0.556 0.579
TfIDF      0.557 0.540

In [200]:
print("DIFFERENCES BY PRECENT:")
((diff_matrix['Now'] - diff_matrix['Before'])*100).sort_values(ascending=False)

DIFFERENCES BY PRECENT:


Sublinear    2.346
Binary       0.217
TfIDF       -1.682
Count       -3.583
dtype: object

# 

#### 5.2.3: GaussianNB

In [201]:
diff_matrix = pd.DataFrame(columns=['Before','Now'],index=['Binary','Count','Sublinear','TfIDF'])

scores_v2.loc['Binary','GussianNB'] = PreformTenFoldCrossValidation(df_train_binary_gnb,GaussianNB())

scores_v2.loc['Count','GussianNB'] = PreformTenFoldCrossValidation(df_train_count_gnb,GaussianNB())

scores_v2.loc['Sublinear','GussianNB'] = PreformTenFoldCrossValidation(df_train_sublinear_gnb,GaussianNB())

scores_v2.loc['TfIDF','GussianNB'] = PreformTenFoldCrossValidation(df_train_tfIDF_gnb,GaussianNB())



diff_matrix['Before'] = scores_v2['GussianNB']
diff_matrix['Now'] = scores['GussianNB']

In [202]:
diff_matrix

Before   Now
Binary     0.599 0.621
Count      0.622 0.639
Sublinear  0.627 0.637
TfIDF      0.640 0.649

In [203]:
print("DIFFERENCES BY PRECENT:")
((diff_matrix['Now'] - diff_matrix['Before'])*100).sort_values(ascending=False)

DIFFERENCES BY PRECENT:


Binary      2.241
Count       1.661
Sublinear   1.042
TfIDF       0.858
dtype: object

# 

#### 5.2.4: MultinomialNB

In [204]:
count_params = {'alpha': 1.0, 'fit_prior': True}

scores_v2.loc['Count','MultinomialNB'] = PreformTenFoldCrossValidation(df_train_count,MultinomialNB(
    alpha=1.0,
    fit_prior=True))

diff_matrix['Before'] = scores['MultinomialNB']
diff_matrix['Now'] = scores_v2['MultinomialNB']

In [205]:
diff_matrix

Before   Now
Binary     0.666   NaN
Count      0.690 0.690
Sublinear  0.667   NaN
TfIDF      0.663   NaN

It can be seen that there is no difference at all in dataset Count

# 

#### 5.2.5: LinearSVC

In [206]:
binary_params = {'C': 0.5, 'loss': 'hinge', 'max_iter': 500}
count_params = {'C': 0.1, 'loss': 'hinge', 'max_iter': 500}
sublinear_params = {'C': 0.5, 'loss': 'hinge', 'max_iter': 500}
tfidf_params = {'C': 0.1, 'loss': 'hinge', 'max_iter': 500}
    
scores_v2.loc['Binary','LinearSVC'] = PreformTenFoldCrossValidation(df_train_binary,LinearSVC(
    C = 0.5,
    loss ='hinge',
    max_iter=500))

scores_v2.loc['Count','LinearSVC'] = PreformTenFoldCrossValidation(df_train_count,LinearSVC(
    C= 0.1,
    loss='hinge',
    max_iter=500))

scores_v2.loc['Sublinear','LinearSVC'] = PreformTenFoldCrossValidation(df_train_sublinear,LinearSVC(
    C = 0.5,
    loss ='hinge',
    max_iter=500))

scores_v2.loc['TfIDF','LinearSVC'] = PreformTenFoldCrossValidation(df_train_tfIDF,LinearSVC(
    C= 0.1,
    loss='hinge',
    max_iter=500))

diff_matrix['Before'] = scores['LinearSVC']
diff_matrix['Now'] = scores_v2['LinearSVC']

In [207]:
diff_matrix

Before   Now
Binary     0.697 0.700
Count      0.660 0.673
Sublinear  0.704 0.697
TfIDF      0.669 0.660

It can be seen that the improvements are not good

# 

#### 5.2.6: Perceptron

In [208]:
binary_params = {'alpha': 0.0001, 'max_iter': 500, 'penalty': None}

scores_v2.loc['Binary','Perceptron'] = PreformTenFoldCrossValidation(df_train_binary,Perceptron(
    alpha=0.0001,
    max_iter=500,
    penalty=None))
                                                                     
diff_matrix['Before'] = scores['Perceptron']
diff_matrix['Now'] = scores_v2['Perceptron']

In [209]:
diff_matrix

Before   Now
Binary     0.679 0.679
Count      0.643   NaN
Sublinear  0.642   NaN
TfIDF      0.633   NaN

No Differences At All

# 

#### 5.2.7: SGDClassifier

In [220]:
binary_params =  {'alpha': 0.01, 'loss': 'hinge', 'max_iter': 500, 'penalty': 'l2'}
count_params = {'alpha': 0.0001, 'loss': 'modified_huber', 'max_iter': 500, 'penalty': 'elasticnet'}
sublinear_params = {'alpha': 0.001, 'loss': 'log', 'max_iter': 500, 'penalty': None}
tfidf_params = {'alpha': 0.0001, 'loss': 'hinge', 'max_iter': 2000, 'penalty': 'l1'}

scores_v2.loc['Binary','SGDClassifier'] = PreformTenFoldCrossValidation(df_train_binary,SGDClassifier(
    alpha=0.01,
    loss='hinge',
    max_iter=500,
    penalty='l2'))

scores_v2.loc['Count','SGDClassifier'] = PreformTenFoldCrossValidation(df_train_count,SGDClassifier(
    alpha=0.0001,
    loss='modified_huber',
    max_iter=500,
    penalty='elasticnet'))

scores_v2.loc['Sublinear','SGDClassifier'] = PreformTenFoldCrossValidation(df_train_sublinear,SGDClassifier(
    alpha=0.001,
    loss='log',
    max_iter=500,
    penalty=None))

scores_v2.loc['TfIDF','SGDClassifier'] = PreformTenFoldCrossValidation(df_train_tfIDF,SGDClassifier(
    alpha=0.0001,
    loss='hinge',
    max_iter=2000,
    penalty='l1'))
                                                                     
diff_matrix['Before'] = scores['SGDClassifier']
diff_matrix['Now'] = scores_v2['SGDClassifier']

In [221]:
diff_matrix

Before   Now
Binary     0.643 0.660
Count      0.651 0.647
Sublinear  0.673 0.646
TfIDF      0.645 0.650

# 

### 5.3: A look at the performance improvement
Note: As mentioned before, NaN values ​​are for sets of algorithms and datasets that we chose not to improve

In [222]:
(scores_v2 - scores) * 100

KNN DecisionTree GussianNB MultinomialNB LinearSVC MLPClassifier  \
Binary    0.000        0.217    -2.241           NaN     0.360           NaN   
Count     0.592       -3.583    -1.661         0.000     1.310           NaN   
Sublinear 7.581        2.346    -1.042           NaN    -0.694           NaN   
TfIDF     6.880       -1.682    -0.858           NaN    -0.928           NaN   

          Perceptron SGDClassifier  
Binary         0.000         1.740  
Count            NaN        -0.404  
Sublinear        NaN        -2.613  
TfIDF            NaN         0.522

## Step 5 Conclusions - What Is The Best Algorithem?
It can be seen that the most significant performance improvements were for KNN

------------
But now, it remains to choose the machine learning algorithm that best suits our problem.
And it is: **LinearSVC**, which gives good results of about 70 percent for all types of vectorizations
------------

# 

## Step 6: Performing the flow on the test set

At this stage we will perform the following flow on the test set:
* We will perform Word Normalization
* We will perform tfIDF Vectorization
* We will run the LinearSVC machine learning algorithm with the hyperparameters we found before
* We will present the first and last 5 predictions

In [261]:

X_Train = df_train[df_train.columns[(df_train.columns!='IsMale') & (df_train.columns!='IsFemale')]]
X_Test = df_test['story']
y_Train = df_train['IsMale']

# 1. Word Normalization
WordNormalize(X_Test)

# 2. tfIDF Vectorization
vec = TfidfVectorizer(use_idf=True,max_features=2000,analyzer='word',min_df=3,max_df=0.8)
vec.fit(list(X_Train['story']))
X_Train_vec = vec.transform(X_Train['story'])
X_Test_vec = vec.transform(X_Test)
X_Train_df = pd.DataFrame(MinMaxScaler().fit_transform(X_Train_vec.toarray()),columns=vec.get_feature_names_out())
X_Test_df = pd.DataFrame(MinMaxScaler().fit_transform(X_Test_vec.toarray()),columns=vec.get_feature_names_out())

# 3. LinearSVC
LSVC = LinearSVC(C=0.1,loss='hinge',max_iter=500)
clf = LSVC.fit(X_Train_df,y_Train)
y_predict = clf.predict(X_Test_df)

# 4. Presenting First 5 And Last 5 Results
print("-------------")
print("FIRST 5: ")
y_predict[:5:]
print("-------------")
print("LAST 5: ")
y_predict[len(y_predict)-5:len(y_predict):]

print("-------------")
df_predicted = pd.DataFrame(columns=['test_example_id','predicted_category'])
df_predicted['test_example_id'] = df_test['test_example_id']
y_predict_ser = pd.Series(y_predict)
df_predicted['predicted_category'] = y_predict_ser.replace({0:'f',1:'m'})
df_predicted

105434

TfidfVectorizer(max_df=0.8, max_features=2000, min_df=3)

-------------
FIRST 5: 


array([1, 1, 1, 0, 0], dtype=int64)

-------------
LAST 5: 


array([1, 1, 1, 1, 1], dtype=int64)

-------------


test_example_id predicted_category
0                  0                  m
1                  1                  m
2                  2                  m
3                  3                  f
4                  4                  f
..               ...                ...
318              318                  m
319              319                  m
320              320                  m
321              321                  m
322              322                  m

[323 rows x 2 columns]

### Save output to csv (optional)
After you're done save your output to the 'classification_results.csv' csv file.<br/>
We assume that the dataframe with your results contain the following columns:
* column 1 (left column): 'test_example_id'  - the same id associated to each of the test stories to be predicted.
* column 2 (right column): 'predicted_category' - the predicted gender value for each of the associated story. 

Assuming your predicted values are in the `df_predicted` dataframe, you should save you're results as following:

In [262]:
df_predicted.to_csv('classification_results.csv',index=False)